File: 2020/four/fourpointtwo.py
#!/user/bin/env python3

from sys import argv
import re

def parse(filename):
  db = []
  with open(filename, 'r') as infile:
    for passport in re.split('\n\n',infile.read().strip()):
      entry = {}
      for record in passport.replace('\n',' ').split(' '):
        key = record.split(':')[0]
        value = record.split(':')[1]
        entry[key] = value
      db.append(entry)
  return (db) 

def inrange(string,minimum,maximum):
  try:
    num = int(string)
  except ValueError:
    return False
  if num >= minimum and num <= maximum:
    return True
  else:
    return False
  
def check(entry):
  fields = ['byr','iyr','eyr','hgt','hcl','ecl','pid']
  if len(entry) < 7:
    return 100 # not enough total fields
  for field in fields:
    if entry.get(field) == None:
      return 200 # missing required fields

  if not inrange(entry['byr'],1920,2002):
    return 300 # birth year out of range
  if not inrange(entry['iyr'],2010,2020):
    return 400 # birth year out of range
  if not inrange(entry['eyr'],2020,2030):
    return 500 # birth year out of range
  if entry['hgt'][-2:] != 'cm' and entry['hgt'][-2:] != 'in':
    return 600 # height not in right format
  if entry['hgt'][-2:] == 'cm':
    if not inrange(entry['hgt'][0:-2],150,193):
      return 650 # height in cm out of range
  elif entry['hgt'][-2:] == 'in':
    if not inrange(entry['hgt'][0:-2],59,76):
      return 700 # height in cm out of range
  if not re.match('^#[0-9a-f]{6}$',entry['hcl']):
    return 800 # hair colour wrong
  cols = ['amb','blu','brn','gry','grn','hzl','oth']
  if entry['ecl'] not in cols:
    return 900 # eye colour wrong
  if len(entry['pid']) != 9:
    return 1000 # passport length wrong
  if not inrange(entry['pid'],0,999999999):
    return 1100 # passport not a number

  if entry.get('cid') == None:
    return 1 # valid but northpole
  return 0 # normal valid

if __name__ == '__main__':
  db = parse(argv[1])
  for entry in db:
    ret = check(entry)
    if ret < 100:
      print ('Valid')
    if ret >= 100:
      print ('Invalid')

    if ret == 0:
      print ('  - Perfect')
    elif ret == 1:
      print ('  - North Pole')
    elif ret == 100:
      print ('  - not enough fields')
    elif ret == 200:
      print ('  - missing required fields')
    elif ret == 300:
      print ('  - birth year out of range')
    elif ret == 400:
      print ('  - issue year out of range')
    elif ret == 500:
      print ('  - expiration year out of range')
    elif ret == 600:
      print ('  - height not cm or in')
    elif ret == 650:
      print ('  - height in cm out of range')
    elif ret == 700:
      print ('  - height in in out of range')
    elif ret == 800:
      print ('  - not a hair colour')
    elif ret == 900:
      print ('  - not a eye colour')
    elif ret == 1000:
      print ('  - passport number not nine')
    elif ret == 1100:
      print ('  - passport number not a number')

File: 2020/four/fourpointtwo.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::collections::HashMap;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
  let file = File::open(filename)?;
  Ok(io::BufReader::new(file).lines())
}

fn parse(filename: &String) -> Vec<HashMap<String,String>> {
  let mut db = Vec::new();
  db.push(HashMap::new());
  let mut entry = db.len()-1;
  if let Ok(lines) = read_lines(filename) {
    for line in lines {
      if let Ok(result) = line {
        for record in result.split(' ') {
          for pair in record.split(':').collect::<Vec<&str>>().chunks(2) {
            if pair.len() == 2 {
              db[entry].insert(pair[0].to_string(),pair[1].to_string());
            } else {
              db.push(HashMap::new());
              entry = db.len()-1;
            }
          }
        }
      }
    }
  }
  return db;
}

fn check(record: &HashMap<String,String>) -> Result<&str,&str>
{
  let fields = ["byr","iyr","eyr","hgt","hcl","ecl","pid"];
  if record.len() < 7 {
    return Err("Not enough total fields");
  }
  for field in &fields {
    if record.get(&field.to_string()) == None {
      return Err("Missing required fields");
    }
  }

  let error_val = "0".to_string();

  let byr = record.get("byr").unwrap_or(&error_val).parse().unwrap_or(0);
  if ! (1920..=2002).contains(&byr) {
    return Err("Birth year out of allowed range"); 
  }
  let iyr = record.get("iyr").unwrap_or(&error_val).parse().unwrap_or(0);
  if ! (2010..=2020).contains(&iyr) {
    return Err("Issue year out of allowed range"); 
  }
  let eyr = record.get("eyr").unwrap_or(&error_val).parse().unwrap_or(0);
  if ! (2020..=2030).contains(&eyr) {
    return Err("Expiration year out of allowed range"); 
  }

  let hgt = record.get("hgt").unwrap_or(&error_val);
  let hgttyp = &hgt[hgt.len()-2 .. ];
  let hgtnum = hgt[..hgt.len()-2].parse().unwrap_or(0);
  if hgttyp != "cm" && hgttyp != "in" {
    return Err("Height not in correct format");
  }
  match hgttyp {
    "cm" => if ! (150..=193).contains(&hgtnum) {
        return Err("Height out of allowed cm range");
      },
    "in" => if ! (59..=76).contains(&hgtnum) {
        return Err("Height out of allowed in range");
      },
    _ => (),
  }

  // should use regex here "^#[0-9a-f]{6}$" but didn't want to use an external crate yet
  let mut count = 0;
  let mut valid = true;
  for chr in record.get("hcl").unwrap_or(&error_val).chars() {
    match count {
      0 => if chr != '#' { valid = false; break; },
      1..=6 => if ! (('a'..='f').contains(&chr) || ('0'..='9').contains(&chr))
        { 
          valid = false; break; 
        },
      _ => { valid = false; break; },
    }
    count += 1;
  }
  if valid == false {
    return Err("Hair colour not valid");
  }

  let cols = ["amb","blu","brn","gry","grn","hzl","oth"];
  let eclclr: &str = &record.get("ecl").unwrap_or(&error_val);
  if ! cols.contains(&eclclr) {
    return Err("Eye colour not valid");
  }

  let pidstr = record.get("pid").unwrap_or(&error_val);
  if pidstr.len() != 9 {
    return Err("Passport number inappropriate length");
  }
  let pid;
  match pidstr.parse() {
    Ok(p) => pid = p,
    Err(_) => return Err("Passport number not a number"),
  }
  if ! (0..=999999999).contains(&pid) {
    return Err("Passport number not valid"); 
  }

  if record.get("cid") == None {
    return Ok("Valid - North Pole");
  }
  return Ok("Valid");
}

fn main() {
  let args: Vec<String> = env::args().collect();
  let db = parse(&args[1]);
  for record in db {
    println!("{:?}",check(&record));
  }
}
